app/src/main/kotlin/compiler/parser/analysis/GrammarAnalysis.kt

```diff
@@ -1,10 +1,41 @@
 package compiler.parser.analysis
 
+import compiler.common.dfa.state_dfa.Dfa
+import compiler.common.dfa.state_dfa.DfaState
 import compiler.parser.grammar.AutomatonGrammar
+import compiler.parser.grammar.Production
 
 class GrammarAnalysis<S : Comparable<S>> {
     fun computeNullable(grammar: AutomatonGrammar<S>): Set<S> {
-        return TODO()
+        val nullable: MutableSet<S> = HashSet()
+        val stateQueue: ArrayDeque<DfaState<S, Production<S>>> = ArrayDeque()
+        val conditionalSets: MutableMap<S, MutableSet<DfaState<S, Production<S>>>> = HashMap()
+
+        grammar.productions.forEach { (_, dfa) -> stateQueue.addAll(dfa.getAcceptingStates()) }
+        fun analyzePredecessors(state: DfaState<S, Production<S>>, dfa: Dfa<S, Production<S>>) {
+            val predecessorsMap = dfa.getPredecessors(state)
+            for ((edge, predecessors) in predecessorsMap) {
+                if (edge in nullable) {
+                    stateQueue.addAll(predecessors)
+                } else {
+                    conditionalSets.getOrPut(edge) { HashSet() }.addAll(predecessors)
+                }
+            }
+        }
+
+        while (stateQueue.size > 0) {
+            val state = stateQueue.removeFirst()
+            grammar.productions.forEach { (_, dfa) -> analyzePredecessors(state, dfa) }
```

Comment:
IIUC this loop only does actually something for one dfa - we can consider adding `getDfa()` method to `DfaState`to avoid looping over all automatons in the grammar  for readability

Reply:
I agree, but I think take care of this after merging this PR.